### PR TITLE
Implementing Dependency injection

### PR DESCRIPTION
**Pull request by iruzevic**
### Added

- Added dependency injection in service locator inside class-main.php
- Changed Manifest.php method to fetch items to non static method.

### Changed

- Updating Manifest.php with new methods

### CODE REVIEW

File: src/assets/class-manifest.php [hunk 1]
```diff
@@ -4,26 +4,90 @@
...
+ * @since 0.7.0 Added Manifest_Data Interface.
+ * @since 0.1.0 Init.
  */
-abstract class Manifest {
+abstract class Manifest implements Service, Manifest_Data {
 
   /**
-   * Provide manifest.json url location.
-   * You project must provide location for the manifest.json for this to work.
+   * Global variable name contant.
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
   * Global variable name constant.
```

File: src/assets/class-manifest.php [hunk 2]
```diff
@@ -4,26 +4,90 @@
...
+   *
+   * @since 0.6.0 Init.
+   */
+  public function register() : void {
+    add_action( 'init', [ $this, 'register_global_variable' ] );
+  }
+
+  /**
+   * Define global variable for assets manifest.
+   * Used to cache data insinde global variable so you don't fetch manifest.json file on every call.
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
   * Used to cache data inside a global variable so you don't have to fetch manifest.json file on every call.
```

File: src/assets/class-manifest.php [hunk 3]
```diff
@@ -4,26 +4,90 @@
...
+   * @return void
+   *
+   * @since 0.6.0 Init.
+   */
+  public function register_global_variable() : void {
+    \define( $this->get_global_variable_name(), $this->get_raw_data() );
+  }
+
+  /**
+   * Return full path for specific asset from manifest.json
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
   * Return full path for specific asset from manifest.json.
```

File: src/assets/class-manifest.php [hunk 4]
```diff
@@ -4,26 +4,90 @@
...
+   */
+  public function get_assets_manifest_item( string $key ) : string {
+    if ( ! $key ) {
+      return '';
+    }
+
+    $data = $this->get_decoded_manifest_data();
+
+    $asset = $data->$key ?? '';
+    if ( empty( $asset ) ) {
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: I'd add space before the conditional and after the conditional. Makes it a bit more readable.

File: src/assets/class-manifest.php [hunk 5]
```diff
@@ -4,26 +4,90 @@
...
+      return '';
+    }
+    return $this->get_assets_manifest_output_prefix() . $asset;
+  }
+
+  /**
+   * Set global variable name to store cached data into.
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: Getter, not setter 🙂

File: src/assets/class-manifest.php [hunk 6]
```diff
@@ -35,14 +99,56 @@ abstract class Manifest {
    *
    * @since 0.1.0
    */
-  public function register_assets_manifest_data() : string {
+  protected function get_raw_data() : string {
 
     $manifest = $this->get_manifest_url();
     if ( ! file_exists( $manifest ) ) {
-      $error_message = esc_html__( 'manifest.json is missing. Bundle the theme before using it.', 'developer-portal' );
+      $error_message = esc_html__( 'manifest.json is missing. Bundle the theme before using it.', 'eightshift-libs' );
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: Add empty line before the conditional please 🙂

File: src/assets/class-manifest.php [hunk 7]
```diff
@@ -35,14 +99,56 @@ abstract class Manifest {
...
       throw Exception\Missing_Manifest::message( $error_message );
     }
 
     return implode( ' ', file( $manifest ) );
   }
+
+  /**
+   * Provide manifest.json url location.
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
   * Get manifest.json url location.
```

File: src/assets/class-manifest.php [hunk 8]
```diff
@@ -35,14 +99,56 @@ abstract class Manifest {
...
+   * If you are using a plugin or a different manifest location provide location with this method.
+   *
+   * @return string
+   *
+   * @since 0.6.0 Changed from abstract to predefined.
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: Predefined what? This since is a bit unclear.

File: src/assets/class-manifest.php [hunk 9]
```diff
@@ -35,14 +99,56 @@ abstract class Manifest {
...
+  /**
+   * Return json_decoded manifest data, now you can call items by object key to get the value.
+   *
+   * @return object Manifest Object.
+   *
+   * @since 0.7.0 Changed to non static method.
+   * @since 0.6.0 Init
+   */
+  protected function get_decoded_manifest_data() {
+    $data = \json_decode( constant( $this->get_global_variable_name() ) );
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: This is neat, didn't use `constant()` before 🙂 👍 

Just one thing, add `true` after the first argument of `json_decode()` function.

Also, feel free to remove all `\` prefixes. This is a php function. Plus I've removed all those from my code, and everything works fine (if function is not found in the current namespace, php will try to find it in namespaces above and finally in the global namespace).

File: src/assets/class-manifest.php [hunk 10]
```diff
@@ -35,14 +99,56 @@ abstract class Manifest {
...
+  /**
+   * Retrun string as a assets output prefix.
+   * Override this if you are using lib for a plugin.
+   *
+   * @return string
+   *
+   * @since 0.6.0
+   */
+  protected function get_assets_manifest_output_prefix() : string {
+    return \home_url( '/' );
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: The name of the method and what it returns is totally confusing to me. Why is here `home_url()` function returned? Also `\` is not needed.

File: src/assets/class-manifest.php [hunk 11]
```diff
@@ -35,14 +99,56 @@ abstract class Manifest {
...
+    $data = \json_decode( constant( $this->get_global_variable_name() ) );
+    if ( ! $data ) {
+      return null;
+    }
+
+    return $data;
+  }
+
+  /**
+   * Retrun string as a assets output prefix.
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
   * Retrun string as an assets output prefix.
```

Can you also give an example of how this looks like in the end, in the comment? I'm not 100% sure what this does and why 🤷‍♂

File: src/assets/interface-manifest-data.php [hunk 12]
```diff
@@ -0,0 +1,26 @@
+<?php
+/**
+ * Interface for Assets.
```
Comment (dingo-d) [inline src/assets/interface-manifest-data.php]:
```suggestion
 * Assets manifest data interface.
 * 
 * Used to define the way manifest item is retrieved from the manifest file.
```

Or something along this line

File: src/class-main.php [hunk 13]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+    return array_map(
+      function( $class ) use ( $container ) {
+        return $container->get( $class );
+      },
+      array_keys( $services )
+    );
+  }
+
+  /**
+   * Get services classes array ad prepare it.
```
Comment (dingo-d) [inline src/class-main.php]:
```suggestion
   * Get services classes array and prepare it for dependency injection.
```

File: src/class-main.php [hunk 14]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+   * Key should be class name, and value empty array or DI items.
```
Comment (dingo-d) [inline src/class-main.php]:
```suggestion
   * Key should be a class name, and value should be an empty array or the dependencies of the class.
```

File: src/class-main.php [hunk 15]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+   *
+   * @return array
+   *
+   * @since 0.7.0 Init.
+   */
+  private function get_service_classes_prepared_array() : array {
+    $output = [];
+    foreach ( $this->get_service_classes() as $key => $value ) {
```
Comment (dingo-d) [inline src/class-main.php]:
```suggestion
    foreach ( $this->get_service_classes() as $class => $dependencies ) {
```

Makes it more clear what you are looping through.

File: src/class-main.php [hunk 16]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+      if ( gettype( $value ) !== 'array' ) {
+        $output[ $value ] = [];
```
Comment (dingo-d) [inline src/class-main.php]: Should this be `$output[ $key ]` (also `$key` == `$class`)?

File: src/class-main.php [hunk 17]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+        $output[ $key ] = $value;
       }
+    }
+
+    return $output;
+  }
+
+  /**
+   * Implement PHP-DI.
+   * Create Object used for PHP-DI to create Dependency Injection methods.
```
Comment (dingo-d) [inline src/class-main.php]:
```suggestion
   * Build and return a DI container.
   * Wire all the dependencies automatically, based on the provided array of class => dependencies from the get_di_items().
```

Something like this.

File: src/class-main.php [hunk 18]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+   *
+   * @param array $services Array of service.
+   * @return object
+   *
+   * @since 0.7.0 Init.
+   */
+  private function get_di_container( array $services ) {
+    $builder = new ContainerBuilder();
+    $builder->useAnnotations( false );
+    $builder->useAutowiring( true );
```
Comment (dingo-d) [inline src/class-main.php]: Actually I think we can safely remove both `$builder->useAnnotations( false );` and `$builder->useAutowiring( true );` as those are defaults. I think I added them out of paranoia 🤣

File: src/class-main.php [hunk 19]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+   * If you pass a class use PHP-DI to prepare if not just output it.
+   *
+   * @param array $items Array of classes/parameters to push in constructor.
+   * @return array
+   *
+   * @since 0.7.0 Init
+   */
+  private function get_di_items( array $items ) : array {
+    return array_map(
+      function( $class ) {
```
Comment (dingo-d) [inline src/class-main.php]: This doesn't have to be a class, no? Maybe `$dependency` would be better? 🙂

File: src/class-main.php [hunk 20]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+  /**
+   * Return prepared Dependency Injection objects.
+   * If you pass a class use PHP-DI to prepare if not just output it.
+   *
+   * @param array $items Array of classes/parameters to push in constructor.
+   * @return array
+   *
+   * @since 0.7.0 Init
+   */
+  private function get_di_items( array $items ) : array {
```
Comment (dingo-d) [inline src/class-main.php]:
```suggestion
  private function get_di_dependencies( array $dependencies ) : array {
```
Semantically makes more sense to me

File: src/class-main.php [hunk 21]
```diff
@@ -77,17 +81,105 @@ public function register_services() : void {
...
+    return array_map(
+      function( $class ) {
+        if ( class_exists( $class ) ) {
+          return \DI\get( $class );
+        }
+        return $class;
+      },
+      $items
```
Comment (dingo-d) [inline src/class-main.php]: If the [above comment](https://github.com/infinum/eightshift-libs/pull/11/files#r289329512) is accepted, don't forget to change this to `$dependencies` as well 🙂

File: src/assets/class-manifest.php (hunk 1 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/assets/class-manifest.php (hunk 4 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/assets/class-manifest.php (hunk 6 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/assets/class-manifest.php (hunk 8 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/assets/class-manifest.php (hunk 9 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/assets/class-manifest.php (hunk 11 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: Manifest gives you and output like this:
```
{
  "application.css": "wp-content/themes/wp-boilerplate/skin/public/application-9d11e4277e8ca74d10ed.css",
  "application.js": "wp-content/themes/wp-boilerplate/skin/public/application.js",
  "applicationAdmin.css": "wp-content/themes/wp-boilerplate/skin/public/applicationAdmin-9d11e4277e8ca74d10ed.css",
  "applicationAdmin.js": "wp-content/themes/wp-boilerplate/skin/public/applicationAdmin.js",
  "jquery.min.js": "wp-content/themes/wp-boilerplate/skin/public/jquery.min.js",
  "logo.svg": "wp-content/themes/wp-boilerplate/skin/public/logo.svg"
}
```

this method adds site url to the final output url.
I have added home_url as a default (using it in theme), if you are using it in plugin you will override this method

File: src/assets/class-manifest.php (hunk 10 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: example in previous comment

File: src/class-main.php (hunk 15 above)
Comment (iruzevic) [reply src/class-main.php]: 👍

File: src/class-main.php (hunk 16 above)
Comment (iruzevic) [reply src/class-main.php]: this is because you can create something like this in you setup:
```
protected function get_service_classes() : array {
    return [
      Assets\Manifest::class,
      Admin\Admin::class => [ Assets\Manifest::class ],
      Admin\Login::class,
      Admin\Media::class,
      Admin\Widgets::class,
      Menu\Menu::class,
      Theme\Theme::class => [ Assets\Manifest::class ],
    ];
  }```

so this array is mixed.
I have looped this array and created it the correct way, so it is easier in the later setup to fetch dependencies.

File: src/class-main.php (hunk 18 above)
Comment (iruzevic) [reply src/class-main.php]: 👍

File: src/class-main.php (hunk 20 above)
Comment (iruzevic) [reply src/class-main.php]: 👍

File: src/class-main.php (hunk 19 above)
Comment (iruzevic) [reply src/class-main.php]: 👍

File: src/class-main.php (hunk 21 above)
Comment (iruzevic) [reply src/class-main.php]: 👍

File: CHANGELOG.md [hunk 22]
```diff
@@ -7,6 +7,19 @@ This projects adheres to [Semantic Versioning](https://semver.org/) and [Keep a
 
 _No documentation available about unreleased changes as of yet._
 
+## [0.7.0]
+
+### Added
+
+- Added dependency injection in service locator inside class-main.php
```
Comment (dingo-d) [inline CHANGELOG.md]: I'd rephrase this. DI and SL are two ways of implementing IoC container, maybe say something like

```
Added DI instead of SL inside the class-main.php
```

File: CHANGELOG.md [hunk 23]
```diff
@@ -7,6 +7,19 @@ This projects adheres to [Semantic Versioning](https://semver.org/) and [Keep a
...
+- Changed Manifest.php method to fetch items to non static method.
```
Comment (dingo-d) [inline CHANGELOG.md]: Same here

```suggestion
- Changed methods used for fetching the manifest items inside Manifest.php to non static methods.
```

File: composer.json [hunk 24]
```diff
@@ -19,7 +19,8 @@
     "source": "https://github.com/infinum/eightshift-libs"
   },
   "require": {
-    "php": ">=7.0"
+    "php": ">=7.0",
```
Comment (dingo-d) [inline composer.json]: Should we up this or leave it as is? 7.0 has reached end of life and is no longer maintained.

https://www.php.net/supported-versions.php

File: src/assets/class-manifest.php [hunk 25]
```diff
@@ -4,26 +4,92 @@
  *
  * It is used to provide manifest.json file location used with Webpack to fetch correct file locations.
  *
+ * @since   0.6.0 Adding multiple methodes for easier extending.
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
 * @since   0.6.0 Added multiple methods for easier extending.
```

File: src/assets/class-manifest.php [hunk 26]
```diff
@@ -4,26 +4,92 @@
...
+  /**
+   * Define global variable for assets manifest.
+   * Used to cache data inside a global variable so you don't have to fetch manifest.json file on every call.
+   *
+   * @return void
+   *
+   * @since 0.6.0 Init.
+   */
+  public function register_global_variable() : void {
+    \define( $this->get_global_variable_name(), $this->get_raw_data() );
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
    define( $this->get_global_variable_name(), $this->get_raw_data() );
```

File: src/assets/class-manifest.php [hunk 27]
```diff
@@ -35,14 +101,58 @@ abstract class Manifest {
...
+   * Get manifest.json url location.
+   * If you are using a plugin or a different manifest location provide location with this method.
+   *
+   * @return string
+   *
+   * @since 0.6.0 Changed from abstract method to prefilled.
+   * @since 0.1.0
+   */
+  protected function get_manifest_url() : string {
+    return get_template_directory() . '/skin/public/manifest.json';
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: Are we planning on using the libs in themes only? If not, then we may need to figure out how to handle this smarter to cover the plugin case.

For my view loading in the plugin I'm writing I have something like

```php
  /**
   * Get the possible locations for the view.
   *
   * @param string $uri URI of the view to get the locations for.
   *
   * @return array Array of possible locations.
   */
  protected function get_locations( $uri ) : array {
    return [
      trailingslashit( \get_stylesheet_directory() ) . $uri,
      trailingslashit( \get_template_directory() ) . $uri,
      trailingslashit( dirname( __DIR__, 2 ) ) . $uri,
    ];
  }```

Then, based on the provided path I check if it's readable (in the method used for validating the uri)

```php
foreach ( $this->get_locations( $uri ) as $location ) {
      if ( is_readable( $location ) ) {
        return $location;
      }
    }
```

Something like this could be done to cover all the cases (themes and plugins.)

File: src/assets/class-manifest.php [hunk 28]
```diff
@@ -35,14 +101,58 @@ abstract class Manifest {
...
+  /**
+   * Return json_decoded manifest data, now you can call items by object key to get the value.
+   *
+   * @return object Manifest Object.
+   *
+   * @since 0.7.0 Changed to non static method.
+   * @since 0.6.0 Init
+   */
+  protected function get_decoded_manifest_data() {
+    $data = \json_decode( constant( $this->get_global_variable_name() ), true );
```
Comment (dingo-d) [inline src/assets/class-manifest.php]:
```suggestion
    $data = json_decode( constant( $this->get_global_variable_name() ), true );
```

File: composer.json (hunk 24 above)
Comment (IvanGrginovInf) [reply composer.json]: If there's a single `public function someMethod() : void {`, this needs to be `>=7.2`

File: src/class-main.php [hunk 29]
```diff
@@ -77,17 +81,104 @@ public function register_services() : void {
...
-    $this->services = array_map(
-      [ $this, 'instantiate_service' ],
-      $classes
-    );
     array_walk(
       $this->services,
-      function( Service $service ) {
-        $service->register();
+      function( $class ) {
+        if ( ! $class instanceof Service ) {
```
Comment (dingo-d) [inline src/class-main.php]: I think I changed this to `Registerable`, because I had one class that just implemented `Registerable` and not `Service`. Just something to think about.

File: src/class-main.php [hunk 30]
```diff
@@ -77,17 +81,104 @@ public function register_services() : void {
...
+   * @param array $services Array of service.
+   * @return object
+   *
+   * @since 0.7.0 Init.
+   */
+  private function get_di_container( array $services ) {
+    $builder = new ContainerBuilder();
+
+    $definitions = [];
+    foreach ( $services as  $key => $values ) {
```
Comment (dingo-d) [inline src/class-main.php]: I'd change `$key` and `$value` into something more descriptive. Also there is an extra space in front of `$key`.

File: CHANGELOG.md (hunk 22 above)
Comment (iruzevic) [reply CHANGELOG.md]: 👍

File: CHANGELOG.md (hunk 23 above)
Comment (iruzevic) [reply CHANGELOG.md]: 👍

File: composer.json (hunk 24 above)
Comment (iruzevic) [reply composer.json]: 👍

File: src/assets/class-manifest.php (hunk 25 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/assets/class-manifest.php (hunk 26 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/assets/class-manifest.php (hunk 28 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: 👍

File: src/class-main.php (hunk 30 above)
Comment (iruzevic) [reply src/class-main.php]: 👍

File: src/assets/class-manifest.php (hunk 27 above)
Comment (iruzevic) [reply src/assets/class-manifest.php]: good idea but I would like to think more about this so I will leave this out for now and add an issue for the future

**Comment by iruzevic** [comment]
all corrections are added. Can you please approve so I can start on the Gutenberg blocks update

File: src/assets/class-manifest.php [hunk 31]
```diff
@@ -4,26 +4,98 @@
...
+    }
+
+    $data = $this->get_decoded_manifest_data();
+
+    if ( ! isset( $data[ $key ] ) ) {
+      $error_message = sprintf(
+        esc_html__( '%s is missing in manifest.json. Please check if provided key is correct.', 'eightshift-libs' ),
+        $key
+      );
+      throw Missing_Manifest::message( $error_message );
```
Comment (dingo-d) [inline src/assets/class-manifest.php]: In case manifest is missing you'd get a type error first, because you'd throw an error, but the expected return type is a `string`. So you can remove the return type hint to fix this (that's what I did whenever I had cases like this).
